### PR TITLE
Rebase353

### PR DESCRIPTION
**Pull request by cbm755**

### CODE REVIEW

File: inst/private/python_header.py [hunk 1]
```diff
@@ -11,16 +11,12 @@
...
-    print(str(e[1]).replace("&", "&amp;").replace("<", "&lt;").replace(">", "&gt;"))
-    print("</f>\n</item>")
-    print("</output_block>\n")
-
+def myerr(error, vital=0):
+    if vital:
+        raise
+    else:
+        octoutput_drv(("INTERNAL_PYTHON_ERROR", error))
+        raise
```
Comment (latot) [inline inst/private/python_header.py]: update this plis, check new code in latests commits

File: inst/private/python_header.py [hunk 2]
```diff
@@ -42,8 +38,7 @@ def myerr(e):
     import xml.etree.ElementTree as ET
     from distutils.version import LooseVersion
 except:
-    myerr(sys.exc_info())
-    raise
+    myerr("Python Header: Error doing imports (perhaps no SymPy?)", 1)
```
Comment (latot) [inline inst/private/python_header.py]: remove ", 1"

File: inst/private/python_header.py [hunk 3]
```diff
@@ -69,8 +64,7 @@ def Version(v):
         # short but not quite right: https://github.com/cbm755/octsympy/pull/320
         return LooseVersion(v.replace('.dev', ''))
 except:
-    myerr(sys.exc_info())
-    raise
+    myerr("Python Header: Error setting functions, block 1", 1)
```
Comment (latot) [inline inst/private/python_header.py]: remove ", 1" plis

File: inst/private/python_header.py [hunk 4]
```diff
@@ -141,9 +135,7 @@ def my_srepr(expr, **settings):
             """return expr in repr form w/ assumptions listed"""
             return _ReprPrinter_w_asm(settings).doprint(expr)
     except:
-        myerr(sys.exc_info())
-        raise
-
+        myerr("Python Header: Error setting functions, block 2", 1)
```
Comment (latot) [inline inst/private/python_header.py]: same

File: inst/private/python_header.py [hunk 5]
```diff
@@ -169,8 +161,7 @@ def octoutput_drv(x):
         else:
             print(DOM.toprettyxml(indent="", newl="\n", encoding="utf-8"))
 except:
-    myerr(sys.exc_info())
-    raise
+    myerr("Python Header: Error setting functions, block 3", 1)
```
Comment (latot) [inline inst/private/python_header.py]: same

File: inst/private/python_header.py [hunk 6]
```diff
@@ -264,8 +255,8 @@ def octoutput(x, et):
             dbout("type: " + str(type(x)))
             octoutput("python does not know how to export type " + str(type(x)), et)
 except:
-    myerr(sys.exc_info())
-    raise
+    myerr("Python Header: Error setting functions, block 4", 1)
```
Comment (latot) [inline inst/private/python_header.py]: same

**Comment by cbm755** [comment]
Thanks.  (I did try to `git cherry-pick` from your branch that commit but there were too many conflicts).

**Comment by latot** [comment]
https://github.com/cbm755/octsympy/pull/358 to merge changes, don't need cherry-pick

**Comment by cbm755** [comment]
how about this?  I haven't tried to make `python_ipc_system` use these ideas yet.

But popen2 should be much safer now.

I put the harded-coded xml function back: previously called `myerr`: I think this should work fine for strings.

**Comment by latot** [comment]
mm, i like some parts and others not, why we can't just call raise from myerr (echo_exception_stdout)?? use it can help to do more easy use this code, and don't call raise everywhere, and don't call octoutput_drv in the myerr stall the system when some error happends D:
i need test some parts because the reason y remove the output of myerr was breaks octave/octsympy

File: inst/private/python_copy_vars_to.m [hunk 7]
```diff
@@ -28,13 +28,13 @@
     %% put inside try-except
     L = do_list(4, in, varargin);
     L = { 'try:' ...
-      sprintf('    %s = []', in) ...
-      L{:} ...
-      '    octoutput_drv("PYTHON: successful variable import")' ...
-      'except:' ...
-      '    octoutput_drv("PYTHON: Error in variable import")' ...
-      '    myerr(sys.exc_info())' ...
```
Comment (cbm755) [inline inst/private/python_copy_vars_to.m]: The double output here is the bit that was causing #352 (and #118).  Basically I just remove one of them...

**Comment by cbm755** [comment]
> why we can't just call raise from myerr (echo_exception_stdout)?? 

Could, I just think the `raise` looks better near the `except:`...  don't feel that strongly...

**Comment by latot** [comment]
okisdokis

**Comment by cbm755** [comment]
Shall I merge this?  Then we can decide if

a)  use your try: `octoutput_drv except: echo_exception_stderr`

b) re-arrange `python_header.py` as you do other places

c) drop the hardcoded xml (you wanted I think).

d) put all of `python_header.py` inside one big `try: except` (instead of blocks).  This is because with popen2, the other blocks get pasted into the stdin even if first one raises an exception...  Maybe I should file another issue about this...

**Comment by latot** [comment]
About d) i'm agree, i think we can order to, vital functions, imports, and others definitions (in that order).

to do that ideas i think should do b and then a
i think can't remove all hardcoded, because a, if fails octoutput_drv we need some way to call.

and the main issue, merge this :D

**Comment by cbm755** [comment]
Ok, fixed the `system` ipc.  Will merge when tests pass.

re: (d)  if just one big try-except block then no reason to re-order: either it all works or it all fails...

**Pull request merged**